arch/arm/include/aarch32/cortex_m/tz_ns.h

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief TrustZone API for use in nonsecure firmware
+ *
+ * TrustZone API for Cortex-M23/M33 CPUs implementing the Security Extension.
+ * The following API can be used by the nonsecure firmware to interact with the
+ * secure firmware.
+ */
+
+#ifndef ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_TZ_NS_H_
+#define ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_TZ_NS_H_
+
+#ifdef _ASMLANGUAGE
+
+/* nothing */
+
+#else
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This macro should be called via @ref __TZ_WRAP_FUNC.
+ *
+ * This macro creates the function body of an "outer" function which behaves
+ * exactly like the wrapped function (@p name), except that the preface function
+ * is called before, and the postface function afterwards.
+ *
+ * @param preface   The function to call first. Must have no parameters and no
+ *                  return value.
+ * @param name      The main function, i.e. the function to wrap. This function
+ *                  will receive the arguments, and its return value will be
+ *                  returned.
+ * @param postface  The function to call last. Must have no parameters and no
+ *                  return value.
+ * @param store_lr  The assembly instruction for storing away the LR value
+ *                  before the functions are called. This instruction must leave
+ *                  r0-r3 unmodified.
+ * @param load_lr   The assembly instruction for restoring the LR value after
+ *                  the functions have been called. This instruction must leave
+ *                  r0-r3 unmodified.
+ */
+#define __TZ_WRAP_FUNC_RAW(preface, name, postface, store_lr, load_lr) \
+	do { \
+		__asm(".global "#preface"; .type "#preface", %function"); \
+		__asm(".global "#name"; .type "#name", %function"); \
+		__asm(".global "#postface"; .type "#postface", %function"); \
+		\
+		__asm(store_lr); \
```

Comment:
You should use `__asm__ volatile` instead of `__asm` as in other inline assembly occurences. I'm not sure now in what cases `__asm` can make troubles but for sure you need `volatile` if you want this piece of code to always appear.